### PR TITLE
The dialogue apache client executor is configurable

### PR DESCRIPTION
**Pull request by carterkozak**
## Before this PR
No way to gather custom metrics 

## After this PR
==COMMIT_MSG==
The dialogue apache client executor is configurable
==COMMIT_MSG==

## Possible downsides?
possible to configure it poorly.

### CODE REVIEW

**Comment by carterkozak** [comment]
I'd like to use a WC instrumented executor when possible.

File: dialogue-apache-hc4-client/src/main/java/com/palantir/dialogue/hc4/ApacheHttpClientChannels.java [hunk 1]
```diff
@@ -235,6 +171,127 @@ public String toString() {
         }
     }
 
+    public static CloseableClientBuilder closeableClientBuilder() {
```
Comment (iamdanfox) [inline dialogue-apache-hc4-client/src/main/java/com/palantir/dialogue/hc4/ApacheHttpClientChannels.java]: maybe just `clientBuilder()` ? or `apacheClientBuilder()` ?

**Comment by iamdanfox** [comment]
Yep I think this makes sense. I should have gone builder from the beginning!

File: dialogue-apache-hc4-client/src/main/java/com/palantir/dialogue/hc4/ApacheHttpClientChannels.java (hunk 1 above)
Comment (carterkozak) [reply dialogue-apache-hc4-client/src/main/java/com/palantir/dialogue/hc4/ApacheHttpClientChannels.java]: good call, I like `clientBuilder()`

File: dialogue-apache-hc4-client/src/main/java/com/palantir/dialogue/hc4/ApacheHttpClientChannels.java [hunk 2]
```diff
@@ -219,9 +150,14 @@ private static MetricName clientPoolSizeMetricName(String clientName, String sta
         private final CloseableHttpClient client;
         private final ResponseLeakDetector leakDetector;
 
-        CloseableClient(CloseableHttpClient client, ResponseLeakDetector leakDetector) {
+        @Nullable
+        private final ExecutorService executor;
```
Comment (iamdanfox) [inline dialogue-apache-hc4-client/src/main/java/com/palantir/dialogue/hc4/ApacheHttpClientChannels.java:152]: possibly worth updating the toString while we're here?

File: dialogue-apache-hc4-client/src/main/java/com/palantir/dialogue/hc4/ApacheHttpClientChannels.java [hunk 3]
```diff
@@ -235,6 +171,127 @@ public String toString() {
...
+            this.clientConfiguration = Preconditions.checkNotNull(value, "ClientConfiguration is required");
+            return this;
+        }
+
+        public CloseableClientBuilder clientName(String value) {
+            this.clientName = Preconditions.checkNotNull(value, "ClientConfiguration is required");
+            return this;
+        }
+
+        public CloseableClientBuilder clientName(ExecutorService value) {
```
Comment (iamdanfox) [inline dialogue-apache-hc4-client/src/main/java/com/palantir/dialogue/hc4/ApacheHttpClientChannels.java]: `s/clientName/executor/`

Also can we describe in javadoc what this executor will get used for? (i.e. a fixedthreadpool with 1 thread is probably a bad idea) is setting up tracing and MDCs beforehand important

File: dialogue-apache-hc4-client/src/main/java/com/palantir/dialogue/hc4/ApacheHttpClientChannels.java (hunk 2 above)
Comment (carterkozak) [reply dialogue-apache-hc4-client/src/main/java/com/palantir/dialogue/hc4/ApacheHttpClientChannels.java:152]: :+1:

File: dialogue-apache-hc4-client/src/main/java/com/palantir/dialogue/hc4/ApacheHttpClientChannels.java (hunk 3 above)
Comment (carterkozak) [reply dialogue-apache-hc4-client/src/main/java/com/palantir/dialogue/hc4/ApacheHttpClientChannels.java]: Updated with docs

File: dialogue-apache-hc4-client/src/main/java/com/palantir/dialogue/hc4/ApacheHttpClientChannels.java [hunk 4]
```diff
@@ -231,7 +164,138 @@ public void close() throws IOException {
...
+
+        private ClientBuilder() {}
+
+        public ClientBuilder clientConfiguration(ClientConfiguration value) {
+            this.clientConfiguration = Preconditions.checkNotNull(value, "ClientConfiguration is required");
+            return this;
+        }
+
+        public ClientBuilder clientName(String value) {
+            this.clientName = Preconditions.checkNotNull(value, "ClientConfiguration is required");
```
Comment (iamdanfox) [inline dialogue-apache-hc4-client/src/main/java/com/palantir/dialogue/hc4/ApacheHttpClientChannels.java]:
```suggestion
            this.clientName = Preconditions.checkNotNull(value, "clientName is required");
```

Comment (iamdanfox) [reply dialogue-apache-hc4-client/src/main/java/com/palantir/dialogue/hc4/ApacheHttpClientChannels.java:197]: I remember also @mattsills was asking about the significance of this field. Perhaps this is also worth some javadoc to say this doesn't influence request behaviour, just used for metrics?

File: dialogue-apache-hc4-client/src/main/java/com/palantir/dialogue/hc4/ApacheHttpClientChannels.java [hunk 5]
```diff
@@ -231,7 +164,138 @@ public void close() throws IOException {
...
+        private ExecutorService executor;
+
+        private ClientBuilder() {}
+
+        public ClientBuilder clientConfiguration(ClientConfiguration value) {
+            this.clientConfiguration = Preconditions.checkNotNull(value, "ClientConfiguration is required");
+            return this;
+        }
+
+        public ClientBuilder clientName(String value) {
```
Comment (iamdanfox) [inline dialogue-apache-hc4-client/src/main/java/com/palantir/dialogue/hc4/ApacheHttpClientChannels.java]: Also I remember @mattsills was asking about the significance of this value. Possibly worth some javadoc explaining this is for metrics and doesn't affect requests?

Also needs to be safe loggable (i.e. people shouldn't put uris in here!)

**Review by iamdanfox — APPROVED** [review]

File: dialogue-apache-hc4-client/src/main/java/com/palantir/dialogue/hc4/ApacheHttpClientChannels.java (hunk 4 above)
Comment (carterkozak) [reply dialogue-apache-hc4-client/src/main/java/com/palantir/dialogue/hc4/ApacheHttpClientChannels.java]: oof, good catch. I'm falling apart!

**Pull request merged**